core/src/locktower.rs

```diff
@@ -0,0 +1,485 @@
+use crate::bank_forks::BankForks;
+use crate::staking_utils;
+use hashbrown::{HashMap, HashSet};
+use solana_runtime::bank::Bank;
+use solana_sdk::account::Account;
+use solana_sdk::pubkey::Pubkey;
+use solana_vote_api::vote_instruction::Vote;
+use solana_vote_api::vote_state::{Lockout, VoteState, MAX_LOCKOUT_HISTORY};
+
+const VOTE_THRESHOLD_DEPTH: usize = 8;
+const VOTE_THRESHOLD_SIZE: f64 = 2f64 / 3f64;
+
+pub struct StakeLockout {
+    lockout: u64,
+    stake: u64,
+}
+
+#[derive(Default)]
+pub struct Locktower {
+    pub max_stake: u64,
```

Comment:
Should this get updated at every epoch?

Reply:
Yea.  @rob-solana, where would be a good spot to do that?

Reply:
the bank's epoch_vote_states() actually returns stakes at the beginning of the epoch

Reply:
we don't want live stakes, right?  we want epoch stakes?

Reply:
```
/// Looks through vote accounts, and finds the latest slot that has achieved
/// supermajority lockout
pub fn get_supermajority_slot(bank: &Bank, epoch_height: u64) -> Option<u64> {
    // Find the amount of stake needed for supermajority
    let stakes_and_lockouts = epoch_stakes_and_lockouts(bank, epoch_height);
    let total_stake: u64 = stakes_and_lockouts.iter().map(|s| s.0).sum();
    let supermajority_stake = total_stake * 2 / 3;

    // Filter out the states that don't have a max lockout
    find_supermajority_slot(supermajority_stake, stakes_and_lockouts.iter())
}
```

Reply:
@rob-solana i just need to know when a new bank is chosen for the epoch so I can update the `Locktower::max_stake` variable.

Reply:
@aeyakovenko, you can ask anytime for max_stake from epoch_vote_accounts()...  you can store a hashmap of epoch->max_stake, then if get(epoch).is_none(), ask the bank and update yourself?

Reply:
I would only want to update this when the leader schedule changes, to the stake size at the schedule.